File: app/src/main/java/com/tokenbrowser/manager/UserManager.java
/*
 * 	Copyright (c) 2017. Token Browser, Inc
 *
 * 	This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation, either version 3 of the License, or
 *     (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License
 *     along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.tokenbrowser.manager;


import android.content.Context;
import android.content.SharedPreferences;

import com.tokenbrowser.crypto.HDWallet;
import com.tokenbrowser.manager.network.IdService;
import com.tokenbrowser.model.local.User;
import com.tokenbrowser.model.network.ServerTime;
import com.tokenbrowser.model.network.UserDetails;
import com.tokenbrowser.util.FileNames;
import com.tokenbrowser.util.FileUtil;
import com.tokenbrowser.util.LogUtil;
import com.tokenbrowser.util.SharedPrefsUtil;
import com.tokenbrowser.view.BaseApplication;

import java.io.File;

import okhttp3.MediaType;
import okhttp3.MultipartBody;
import okhttp3.RequestBody;
import retrofit2.adapter.rxjava.HttpException;
import rx.Single;
import rx.schedulers.Schedulers;
import rx.subjects.BehaviorSubject;

public class UserManager {

    private final static String FORM_DATA_NAME = "Profile-Image-Upload";
    private final static String OLD_USER_ID = "uid";
    private final static String USER_ID = "uid_v2";


    private final BehaviorSubject<User> userSubject = BehaviorSubject.create();
    private SharedPreferences prefs;
    private HDWallet wallet;

    /* package */ UserManager() {}

    public final BehaviorSubject<User> getUserObservable() {
        return this.userSubject;
    }

    public final Single<User> getCurrentUser() {
        return
                this.userSubject
                .filter(user -> user != null)
                .first()
                .toSingle()
                .doOnError(t -> LogUtil.exception(getClass(), "getCurrentUser", t))
                .onErrorReturn(null);
    }

    public UserManager init(final HDWallet wallet) {
        this.wallet = wallet;
        this.prefs = BaseApplication.get().getSharedPreferences(FileNames.USER_PREFS, Context.MODE_PRIVATE);
        attachConnectivityListener();
        return this;
    }

    private void attachConnectivityListener() {
        // Whenever the network changes init the user.
        // This is dumb and potentially inefficient but it shouldn't have
        // any adverse effects and it is easy to improve later.
        BaseApplication
                .get()
                .isConnectedSubject()
                .subscribe(
                        isConnected -> initUser(),
                        this::handleUserError
                );
    }

    private void initUser() {
        if (userNeedsToRegister()) {
            registerNewUser();
        } else if (userNeedsToMigrate()) {
            migrateUser();
        } else if (SharedPrefsUtil.shouldForceUserUpdate()) {
            forceUpdateUser();
        } else {
            getExistingUser();
        }
    }

    private boolean userNeedsToRegister() {
        final String oldUserId = this.prefs.getString(OLD_USER_ID, null);
        final String newUserId = this.prefs.getString(USER_ID, null);
        final String expectedAddress = this.wallet.getOwnerAddress();
        final String userId = newUserId == null ? oldUserId : newUserId;
        return userId == null || !userId.equals(expectedAddress);
    }

    private boolean userNeedsToMigrate() {
        final String userId = this.prefs.getString(USER_ID, null);
        final String expectedAddress = this.wallet.getOwnerAddress();
        return userId == null || !userId.equals(expectedAddress);
    }

    private void registerNewUser() {
        getTimestamp()
        .subscribe(
                this::registerNewUserWithTimestamp,
                this::handleUserError
        );
    }

    private void handleUserError(final Throwable throwable) {
        LogUtil.exception(getClass(), "Unable to register/fetch user", throwable);
    }

    private void registerNewUserWithTimestamp(final ServerTime serverTime) {
        final UserDetails ud = new UserDetails().setPaymentAddress(this.wallet.getPaymentAddress());

        IdService
            .getApi()
            .registerUser(ud, serverTime.get())
            .subscribe(
                    this::updateCurrentUser,
                    this::handleUserRegistrationFailed
            );
        SharedPrefsUtil.setForceUserUpdate(false);
    }

    private void handleUserRegistrationFailed(final Throwable throwable) {
        LogUtil.error(getClass(), throwable.toString());
        if (throwable instanceof HttpException && ((HttpException)throwable).code() == 400) {
            getExistingUser();
        }
    }

    private void getExistingUser() {
        IdService
            .getApi()
            .getUser(this.wallet.getOwnerAddress())
            .subscribe(
                    this::updateCurrentUser,
                    this::handleUserError
            );
    }

    private void updateCurrentUser(final User user) {
        prefs
            .edit()
            .putString(USER_ID, user.getTokenId())
            .apply();
        this.userSubject.onNext(user);
    }

    private void migrateUser() {
        forceUpdateUser();
        SharedPrefsUtil.setWasMigrated(true);
    }

    private void forceUpdateUser() {
        final UserDetails ud = new UserDetails().setPaymentAddress(this.wallet.getPaymentAddress());
        updateUser(ud).subscribe(
                __ -> {},
                this::handleUserError
        );
        SharedPrefsUtil.setForceUserUpdate(false);
    }

    public Single<User> updateUser(final UserDetails userDetails) {
        return getTimestamp()
                .subscribeOn(Schedulers.io())
                .flatMap(serverTime -> updateUserWithTimestamp(userDetails, serverTime));
    }

    private Single<User> updateUserWithTimestamp(
            final UserDetails userDetails,
            final ServerTime serverTime) {

        return IdService.getApi()
                .updateUser(this.wallet.getOwnerAddress(), userDetails, serverTime.get())
                .subscribeOn(Schedulers.io())
                .doOnSuccess(this::updateCurrentUser);
    }

    public Single<User> uploadAvatar(final File file) {
        final FileUtil fileUtil = new FileUtil();
        final String mimeType = fileUtil.getMimeTypeFromFilename(file.getName());
        if (mimeType == null) {
            return Single.error(new IllegalArgumentException("Unable to determine file type from file."));
        }
        final MediaType mediaType = MediaType.parse(mimeType);
        final RequestBody requestFile = RequestBody.create(mediaType, file);
        final MultipartBody.Part body = MultipartBody.Part.createFormData(FORM_DATA_NAME, file.getName(), requestFile);

        return getTimestamp()
                .subscribeOn(Schedulers.io())
                .flatMap(serverTime ->
                        IdService
                        .getApi()
                        .uploadFile(body, serverTime.get()))
                .doOnSuccess(this.userSubject::onNext);
    }

    private Single<ServerTime> getTimestamp() {
        return IdService
                .getApi()
                .getTimestamp();
    }

    public Single<List<User>> getTopRatedPublicUsers(final int limit) {
        return getTimestamp()
                .flatMap(serverTime -> getTopRatedPublicUsers(serverTime, limit))
                .map(UserSearchResults::getResults)
                .subscribeOn(Schedulers.io());
    }

    public Single<List<User>> getLatestPublicUsers(final int limit) {
        return getTimestamp()
                .flatMap(serverTime -> getLatestPublicUsers(serverTime, limit))
                .map(UserSearchResults::getResults)
                .subscribeOn(Schedulers.io());
    }

    private Single<UserSearchResults> getTopRatedPublicUsers(final ServerTime serverTime,
                                                             final int limit) {
        return IdService
                .getApi()
                .getUsers(true, true, false, limit, serverTime.get());
    }

    private Single<UserSearchResults> getLatestPublicUsers(final ServerTime serverTime,
                                                           final int limit) {
        return IdService
                .getApi()
                .getUsers(true, false, true, limit, serverTime.get());
    }

    public Single<Void> webLogin(final String loginToken) {
        return getTimestamp()
                .subscribeOn(Schedulers.io())
                .observeOn(Schedulers.io())
                .flatMap((serverTime) -> webLoginWithTimestamp(loginToken, serverTime));
    }

    private Single<Void> webLoginWithTimestamp(final String loginToken, final ServerTime serverTime) {
        if (serverTime == null) {
            throw new IllegalStateException("ServerTime was null");
        }

        return IdService
                .getApi()
                .webLogin(loginToken, serverTime.get());
    }

    public void clear() {
        this.prefs
                .edit()
                .putString(USER_ID, null)
                .apply();
        this.userSubject.onNext(null);
    }
}

File: app/src/main/java/com/toshi/manager/UserManager.kt
/*
 * 	Copyright (c) 2017. Toshi Inc
 *
 * 	This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation, either version 3 of the License, or
 *     (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License
 *     along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.toshi.manager

import com.toshi.crypto.HDWallet
import com.toshi.manager.network.IdService
import com.toshi.model.local.User
import com.toshi.model.network.ServerTime
import com.toshi.model.network.UserDetails
import com.toshi.util.logging.LogUtil
import com.toshi.util.sharedPrefs.AppPrefs
import com.toshi.util.sharedPrefs.AppPrefsInterface
import com.toshi.util.sharedPrefs.UserPrefs
import com.toshi.util.sharedPrefs.UserPrefsInterface
import com.toshi.util.uploader.FileUploader
import com.toshi.view.BaseApplication
import retrofit2.HttpException
import rx.Completable
import rx.Observable
import rx.Scheduler
import rx.Single
import rx.Subscription
import rx.schedulers.Schedulers
import rx.subjects.BehaviorSubject
import rx.subscriptions.CompositeSubscription
import java.io.File
import java.util.concurrent.TimeUnit

class UserManager(
        private val idService: IdService = IdService.get(),
        private val fileUploader: FileUploader = FileUploader(idService.api),
        private val userPrefs: UserPrefsInterface = UserPrefs(),
        private val appPrefs: AppPrefsInterface = AppPrefs,
        private val recipientManager: RecipientManager,
        private val baseApplication: BaseApplication = BaseApplication.get(),
        private val scheduler: Scheduler = Schedulers.io()
) {

    private val isConnectedSubject by lazy { baseApplication.isConnectedSubject }

    private val subscriptions by lazy { CompositeSubscription() }
    private val userSubject by lazy { BehaviorSubject.create<User>() }
    private var connectivitySub: Subscription? = null
    private var wallet: HDWallet? = null

    fun init(wallet: HDWallet): Completable {
        this.wallet = wallet
        attachConnectivityListener()
        return initUser()
    }

    private fun attachConnectivityListener() {
        // Whenever the network changes init the user.
        // This is dumb and potentially inefficient but it shouldn't have
        // any adverse effects and it is easy to improve later.
        clearConnectivitySubscription()
        connectivitySub = isConnectedSubject
                .skip(1) // Skip the cached value in the subject.
                .subscribe(
                        { handleConnectivity(it) },
                        { LogUtil.exception("Error while initiating user $it") }
                )
    }

    private fun handleConnectivity(isConnected: Boolean) {
        if (isConnected) initUser()
                .subscribeOn(scheduler)
                .subscribe({}, {})
    }

    private fun initUser(): Completable {
        return when {
            userNeedsToRegister() -> registerNewUser()
            userNeedsToMigrate() -> migrateUser()
            appPrefs.shouldForceUserUpdate() -> forceUpdateUser()
            else -> fetchAndUpdateUser()
        }
    }

    private fun userNeedsToRegister(): Boolean {
        val oldUserId = userPrefs.getOldUserId()
        val newUserId = userPrefs.getUserId()
        val expectedAddress = wallet?.ownerAddress
        val userId = newUserId ?: oldUserId
        return userId == null || userId != expectedAddress
    }

    private fun userNeedsToMigrate(): Boolean {
        val userId = userPrefs.getUserId()
        val expectedAddress = wallet?.ownerAddress
        return userId == null || userId != expectedAddress
    }

    private fun registerNewUser(): Completable {
        return getTimestamp()
                .flatMap { registerNewUserWithTimestamp(it) }
                .onErrorResumeNext { handleUserRegistrationFailed(it) } // If the user is already registered, the server will give a 400 response.
                .doOnSuccess { updateCurrentUser(it) }
                .doOnError { LogUtil.exception("Error while registering user with timestamp") }
                .toCompletable()
    }

    private fun registerNewUserWithTimestamp(serverTime: ServerTime): Single<User> {
        val wallet = wallet ?: return Single.error(IllegalStateException("Wallet is null while registerNewUserWithTimestamp"))
        val userDetails = UserDetails(payment_address = wallet.paymentAddress)
        appPrefs.setForceUserUpdate(false)
        return idService.api.registerUser(userDetails, serverTime.get())
    }

    private fun handleUserRegistrationFailed(throwable: Throwable): Single<User> {
        return if (throwable is HttpException && throwable.code() == 400) forceFetchUserFromNetworkSingle()
        else Single.error(throwable)
    }

    fun getCurrentUserObservable(): Observable<User> {
        forceFetchUserFromNetwork()
        return userSubject.asObservable()
    }

    private fun forceFetchUserFromNetworkSingle(): Single<User> {
        return getWallet()
                .flatMap { idService.api.forceGetUser(it.ownerAddress) }
                .doOnSuccess { updateCurrentUser(it) }
                .doOnError { LogUtil.exception("Error while fetching user from network $it") }
    }

    private fun forceFetchUserFromNetwork() {
        val sub = getWallet()
                .flatMap { idService.api.forceGetUser(it.ownerAddress) }
                .subscribe(
                        { updateCurrentUser(it) },
                        { LogUtil.exception("Error while fetching user from network $it") }
                )

        subscriptions.add(sub)
    }

    private fun fetchAndUpdateUser(): Completable {
        return getWallet()
                .flatMap { recipientManager.getUserFromPaymentAddress(it.paymentAddress) }
                .doOnSuccess { updateCurrentUser(it) }
                .doOnError { LogUtil.exception("Error while fetching user from network $it") }
                .toCompletable()
                .onErrorComplete()
    }

    private fun updateCurrentUser(user: User) {
        userPrefs.setUserId(user.toshiId)
        userSubject.onNext(user)
        recipientManager.cacheUser(user)
    }

    private fun migrateUser(): Completable {
        appPrefs.setWasMigrated(true)
        return forceUpdateUser()
    }

    private fun forceUpdateUser(): Completable {
        return getWallet()
                .map { UserDetails(payment_address = it.paymentAddress) }
                .flatMap { updateUser(it) }
                .doOnSuccess { AppPrefs.setForceUserUpdate(false) }
                .doOnError { LogUtil.exception("Error while updating user while initiating $it") }
                .toCompletable()
                .onErrorComplete()
    }

    fun updateUser(userDetails: UserDetails): Single<User> {
        return getTimestamp()
                .flatMap { updateUserWithTimestamp(userDetails, it) }
                .subscribeOn(scheduler)
    }

    private fun updateUserWithTimestamp(userDetails: UserDetails, serverTime: ServerTime): Single<User> {
        return getWallet()
                .flatMap { idService.api.updateUser(it.ownerAddress, userDetails, serverTime.get()) }
                .subscribeOn(scheduler)
                .doOnSuccess { updateCurrentUser(it) }
    }

    fun uploadAvatar(file: File): Single<User> {
        return fileUploader
                .uploadAvatar(file)
                .subscribeOn(scheduler)
                .doOnSuccess { userSubject.onNext(it) }
    }

    private fun getTimestamp() = idService.api.timestamp

    fun webLogin(loginToken: String): Completable {
        return getTimestamp()
                .flatMapCompletable { webLoginWithTimestamp(loginToken, it) }
                .subscribeOn(scheduler)
    }

    private fun webLoginWithTimestamp(loginToken: String, serverTime: ServerTime?): Completable {
        if (serverTime == null) throw IllegalStateException("ServerTime was null")
        return idService.api.webLogin(loginToken, serverTime.get())
    }

    fun getUserObservable(): Observable<User> = userSubject.asObservable()

    fun getCurrentUser(): Single<User?> {
        return userSubject
                .first()
                .toSingle()
                .timeout(20, TimeUnit.SECONDS)
                .doOnError { LogUtil.exception("getCurrentUser $it") }
                .onErrorReturn { null }
    }

    private fun getWallet(): Single<HDWallet> {
        return Single.fromCallable {
            while (wallet == null) Thread.sleep(100)
            return@fromCallable wallet ?: throw IllegalStateException("Wallet is null UserManager::getWallet")
        }
        .subscribeOn(scheduler)
        .timeout(20, TimeUnit.SECONDS)
    }

    fun clear() {
        subscriptions.clear()
        clearConnectivitySubscription()
        userPrefs.clear()
        userSubject.onNext(null)
    }

    private fun clearConnectivitySubscription() = connectivitySub?.unsubscribe()
}